pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/instance/InstanceAssignmentDriver.java

```diff
@@ -77,7 +95,31 @@ public InstancePartitions assignInstances(InstancePartitionsType instancePartiti
         new InstanceReplicaGroupPartitionSelector(assignmentConfig.getReplicaGroupPartitionConfig(), tableNameWithType);
     InstancePartitions instancePartitions = new InstancePartitions(
         instancePartitionsType.getInstancePartitionsName(TableNameBuilder.extractRawTableName(tableNameWithType)));
+    if (shouldRetainInstanceSequence) {
+      // Keep the pool to instances map if instance sequence should be retained.
+      instancePartitions
+          .setPoolToInstancesMap(extractInstanceNamesFromPoolToInstanceConfigsMap(poolToInstanceConfigsMap));
+    }
```

Comment:
Does this mean if we do not set `shouldRetainInstanceSequence` in the initial assignment, the subsequent table rebalance will not be able to retain sequence even if we set `retainInstanceSequence` in the rebalance config? Is it okay that we always persist the instance sequence in the initial assignment?

Reply:
IMO we should only persist the instance sequence if it's required. 

The actual value of `shouldRetainInstanceSequence` is determined by  `(existingPoolToInstancesMap != null)`. There are two cases need to differentiate:
* If it's false, that means `retainInstanceSequence` is set to false in the rebalance config. There is no need to retain instance sequence. 
* If it's true but the map is empty, that means `retainInstanceSequence` is set to true but there is no such map in the ZK yet.

Inside the driver, if `retainInstanceSequence` from rebalance config is true, the existingPoolToInstancesMap should be fetched from ZK. The first fetch would return empty map since there is no such map persisted in the ZK yet. But since `shouldRetainInstanceSequence` is true because `(existingPoolToInstancesMap != null)`, this new map will be persisted into the ZK for the next calculation. So next time, when the same request comes, this map can be reused.

Reply:
I'm wondering would it be better if we can have one config instead of two to avoid some confusion. I mean as a user I might expect the `retainInstanceSequence` in rebalance to retain current sequence, whereas it's actually a "best effort" depending on if the corresponding flag is set in the initial assignment. 

Similarly for current large use-cases without this map already available we might need some migration plan?

Reply:
Good question! Although there are two added configs, they actually are not in the same method. `retainInstanceSequence` is in the caller method of `InstanceAssignmentDriver.assignInstances()`.
`shouldRetainInstanceSequence` is inside the method of `InstanceAssignmentDriver.assignInstances()`.
these two configs denotes the same thing, i.e. whether to denote whether this feature should be enabled or not.